src/client/libds3/bucket.c

```diff
@@ -7,31 +7,343 @@
+		return -ENOMEM;
+	}
+
+	// TODO: Handle markers and other bucket info
+	rc = daos_pool_list_cont(ds3->poh, &ncont, conts, ev);
+	if (rc == 0) {
+		*is_truncated = false;
+	} else if (rc == -DER_TRUNC) {
+		rc            = 0;
+		*is_truncated = true;
+	} else {
+		D_ERROR("Failed to list containers in pool, rc = %d\n", rc);
+		rc = daos_der2errno(rc);
+		goto err;
+	}
+
+	daos_size_t   bi   = 0;
+	ds3_bucket_t *ds3b = NULL;
+	for (int i = 0; i < ncont; i++) {
+		char *name = conts[i].pci_label;
+		if (strcmp(name, METADATA_BUCKET) == 0) {
+			D_INFO("Skipping _METADATA bucket");
+			continue;
+		}
+
+		// Copy bucket name
+		strcpy(buf[bi].name, name);
+
+		// Get info
+		rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+		if (rc)
+			goto err;
+
+		rc = ds3_bucket_get_info(&buf[bi], ds3b, ev);
+
+		ds3_bucket_close(ds3b, ev);
+
+		if (rc)
+			goto err;
+
+		bi++;
+		// TODO handle marker
+	}
+
+	*nbuck = bi;
+
+err:
+	D_FREE(conts);
+	return -rc;
 }
 
 int
-ds3_bucket_create(const char *name, daos_prop_t *props, ds3_t *ds3, daos_event_t *ev)
+ds3_bucket_create(const char *name, struct ds3_bucket_info *info, dfs_attr_t *attr, ds3_t *ds3,
+		  daos_event_t *ev)
 {
-	return 0;
+	if (ds3 == NULL || name == NULL)
+		return -EINVAL;
+
+	// Prevent attempting to create metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot create metadata bucket");
+		return -EINVAL;
+	}
+
+	int rc = 0;
+
+	// Create dfs container and open ds3b
+	rc = dfs_cont_create_with_label(ds3->poh, name, attr, NULL, NULL, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to create container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	ds3_bucket_t *ds3b = NULL;
+	rc                 = ds3_bucket_open(name, &ds3b, ds3, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to open container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	rc = ds3_bucket_set_info(info, ds3b, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to put bucket info, rc = %d\n", rc);
+		goto err;
+	}
+
+	// Create multipart index
+	rc = dfs_mkdir(ds3->meta_dfs, ds3->meta_dirs[MULTIPART_DIR], name, DEFFILEMODE, 0);
+	if (rc != 0 && rc != EEXIST) {
+		D_ERROR("Failed to create multipart index, rc = %d\n", rc);
+	}
+
+err:
+	ds3_bucket_close(ds3b, ev);
+	return -rc;
 }
 
 int
 ds3_bucket_destroy(const char *name, ds3_t *ds3, daos_event_t *ev)
 {
+	// Prevent attempting to destroy metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot destroy metadata bucket");
+		return -ENOENT;
+	}
+
 	return 0;
 }
 
 int
 ds3_bucket_open(const char *name, ds3_bucket_t **ds3b, ds3_t *ds3, daos_event_t *ev)
 {
+	if (ds3 == NULL || name == NULL || ds3b == NULL)
+		return -EINVAL;
+
+	// Prevent attempting to open metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot open metadata bucket");
+		return -ENOENT;
+	}
+
+	int           rc;
+	ds3_bucket_t *ds3b_tmp;
+	D_ALLOC_PTR(ds3b_tmp);
+	if (ds3b_tmp == NULL)
+		return -ENOMEM;
+
+	// TODO: We need to cache open container handles
```

Comment:
this is pretty important actually as cont open is pretty expensive.
have you considered using the dfs_connect API which does this caching for you?

Reply:
Yes, but I wish it also returned `coh`, or there was a way to access `coh` from `dfs`. We need `coh` for `daos_cont_get_attr`.

Reply:
dfs_cont_get/put seems to be what you are looking for. I agree with @mchaarawi that we should use dfs_connect()/disconnect().

Reply:
Done